Imperative invokers for popover don't fix focus order

lukewarlow opened the pull request:
#### b46f59f1c06a8d2f0562fee7c42472c99d7dee85
<pre>
Imperative invokers for popover don&apos;t fix focus order
<a href="https://bugs.webkit.org/show_bug.cgi?id=286575">https://bugs.webkit.org/show_bug.cgi?id=286575</a>

Reviewed by Tim Nguyen.

The FocusController currently still relies on the popoverTargetElement of buttons, to look for popovers
to move to the new focus order.

This patch stores the invoked popover on the provided invoking element&apos;s rare data,
and then updates FocusController to use this instead.

This way both declarative and imperative popover relationships correctly fix focus order.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2.html:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::invokedPopover const):
(WebCore::Element::setInvokedPopover):
* Source/WebCore/dom/Element.h:
(WebCore::Element::handleCommandInternal):
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::invokedPopover const):
(WebCore::ElementRareData::setInvokedPopover):
(WebCore::ElementRareData::useTypes const):
* Source/WebCore/dom/NodeRareData.h:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::handleCommandInternal):
* Source/WebCore/html/HTMLDialogElement.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::showPopoverInternal):
(WebCore::HTMLElement::setInvoker):
(WebCore::HTMLElement::hidePopoverInternal):
(WebCore::HTMLElement::handleCommandInternal):
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::handlePopoverTargetAction):
(WebCore::HTMLFormControlElement::handlePopoverTargetAction const): Deleted.
* Source/WebCore/html/HTMLFormControlElement.h:
* Source/WebCore/page/FocusController.cpp:
(WebCore::openPopoverForInvoker):

Canonical link: <a href="https://commits.webkit.org/291719@main">https://commits.webkit.org/291719@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0cb277c6bfd623d1aca25f3e4f2416556c7e6b71

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93745 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13319 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3056 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98751 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44271 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95795 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13614 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21761 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71567 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28942 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96747 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10138 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84730 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51901 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9819 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2358 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43586 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80092 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2426 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100786 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20797 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15177 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80591 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21049 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80673 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79931 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19898 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24471 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1829 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13954 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20781 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/25959 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20468 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23928 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22209 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-3.html:
```diff
@@ -0,0 +1,244 @@
+<!DOCTYPE html>
```
Exported at https://github.com/web-platform-tests/wpt/pull/50308

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1098,6 +1098,9 @@ ExceptionOr<void> HTMLElement::showPopoverInternal(const HTMLElement* invoker)
         popoverData()->setInvoker(invoker);
 
+    if (invoker)
+        invoker->setInvokedPopover(this);
```
I feel like this should be called from `PopoverData::setInvoker` and that method should be moved to the C++ file.

lukewarlow replied:
We don't have access to the popover element from within PopoverData though do we?

lukewarlow replied:
I guess we can update setInvoker() to take both the popover and the invoker to do this?

lukewarlow replied:
I've moved it in there, hopefully it's as you'd like.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-3.html:
Could you restructure popover-focus-2 in a way that tests both variants?

I've added a subtest recently and it's not reflected over in this test.

lukewarlow replied:
Done

rniwa commented on Source/WebCore/html/HTMLElement.h:
```diff
@@ -212,6 +215,8 @@ class HTMLElement : public StyledElement {
 
     void mapLanguageAttributeToLocale(const AtomString&, MutableStyleProperties&);
 
+    WeakPtr<HTMLElement, WeakPtrImplWithEventTargetData> m_invokedPopover;
```
This should be in ElementRareData or on a side table.

rniwa commented on Source/WebCore/dom/ElementRareData.h:
```diff
@@ -275,6 +282,8 @@ class ElementRareData : public NodeRareData {
 
     std::unique_ptr<PopoverData> m_popoverData;
 
+    RefPtr<Element> m_invokedPopover;
```
Shouldn't this be a WeakPtr???

lukewarlow replied:
Yes it probably should be I've marked this as draft for now I need to work it how to do that correctly. It seemed to mess with the rare data size when I tried before.

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3087,6 +3087,17 @@ void Element::clearPopoverData()
+
+void Element::setInvokedPopover(RefPtr<Element>&& element)
+{
+    ElementRareData& data = ensureElementRareData();
```
```suggestion
    auto& data = ensureElementRareData();
```

nt1m commented on Source/WebCore/dom/PopoverData.cpp:
```diff
@@ -42,4 +42,13 @@ Ref<ToggleEventTask> PopoverData::ensureToggleEventTask(Element& element)
+        oldInvoker->setInvokedPopover(nullptr);
+    m_invoker = invoker;
+    if (RefPtr<Element> newInvoker = m_invoker.get())
+        newInvoker->setInvokedPopover(popover ? RefPtr<Element>(popover) : nullptr);
```
```suggestion
    if (RefPtr oldInvoker = m_invoker.get(); oldInvoker && !invoker)
        oldInvoker->setInvokedPopover(nullptr);
    m_invoker = invoker;
    if (RefPtr newInvoker = m_invoker.get())
        newInvoker->setInvokedPopover(popover ? RefPtr { popover } : nullptr);
```

nt1m commented on Source/WebCore/dom/PopoverData.cpp:
```diff
@@ -42,4 +42,13 @@ Ref<ToggleEventTask> PopoverData::ensureToggleEventTask(Element& element)
+    m_invoker = invoker;
+    if (RefPtr newInvoker = m_invoker.get())
+        newInvoker->setInvokedPopover(popover ? RefPtr { popover } : nullptr);
+}
```
Passing both the popover and the invoker to this function is kinda yucky. @rniwa suggested offline we could keep make this method only do `m_invoker = invoker;` and create a new `HTMLElement::setInvoker()` method that sets the invoker & does the extra logic of changing `setInvokedPopover` and use that at the call sites.

We should also make `PopoverData::setInvoker` private/protected so people don't call the wrong method.

lukewarlow replied:
PopoverData::setInvoker can't be private or protected because else it cant be called inside HTMLElement::setInvoker.
Otherwise I've done the suggested changes.

nt1m reviewed (commented):
This looks generally fine aside from one comment.

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1160,6 +1160,15 @@ ExceptionOr<void> HTMLElement::showPopoverInternal(const HTMLElement* invoker)
     return { };
 }
 
+void HTMLElement::setInvoker(HTMLElement* popover, HTMLElement* invoker)
```
You shouldn't need to pass the popover. `this` is the popover itself.

lukewarlow replied:
in hidePopoverInternal we call this with nullptr, nullptr but I think you're right this is trying to handle a case that doesn't happen. Will double check quickly, but if it works I'll drop that param and push up soon.

lukewarlow replied:
Fixed.

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1160,6 +1160,15 @@ ExceptionOr<void> HTMLElement::showPopoverInternal(const HTMLElement* invoker)
 
+void HTMLElement::setInvoker(HTMLElement* invoker)
+{
+    if (RefPtr oldInvoker = popoverData()->invoker(); oldInvoker && !invoker)
```
I'm not sure I understand why `!invoker` is correct here. If it is, please add a comment, if not, please remove.

lukewarlow replied:
I believe it was in one iteration and got left over but not needed now.

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1160,6 +1160,15 @@ ExceptionOr<void> HTMLElement::showPopoverInternal(const HTMLElement* invoker)
 
+void HTMLElement::setInvoker(HTMLElement* invoker)
+{
+    if (RefPtr oldInvoker = popoverData()->invoker(); oldInvoker)
```
```suggestion
    if (RefPtr oldInvoker = popoverData()->invoker())
```

lukewarlow committed:
Imperative invokers for popover don't fix focus order
https://bugs.webkit.org/show_bug.cgi?id=286575

Reviewed by Tim Nguyen.

The FocusController currently still relies on the popoverTargetElement of buttons, to look for popovers
to move to the new focus order.

This patch stores the invoked popover on the provided invoking element's rare data,
and then updates FocusController to use this instead.

This way both declarative and imperative popover relationships correctly fix focus order.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2.html:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::invokedPopover const):
(WebCore::Element::setInvokedPopover):
* Source/WebCore/dom/Element.h:
(WebCore::Element::handleCommandInternal):
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::invokedPopover const):
(WebCore::ElementRareData::setInvokedPopover):
(WebCore::ElementRareData::useTypes const):
* Source/WebCore/dom/NodeRareData.h:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::handleCommandInternal):
* Source/WebCore/html/HTMLDialogElement.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::showPopoverInternal):
(WebCore::HTMLElement::setInvoker):
(WebCore::HTMLElement::hidePopoverInternal):
(WebCore::HTMLElement::handleCommandInternal):
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::handlePopoverTargetAction):
(WebCore::HTMLFormControlElement::handlePopoverTargetAction const): Deleted.
* Source/WebCore/html/HTMLFormControlElement.h:
* Source/WebCore/page/FocusController.cpp:
(WebCore::openPopoverForInvoker):

Canonical link: https://commits.webkit.org/291719@main
